### ISSUE TITLE
Replace Flow with TypeScript to avoid Facebook’s BSD Patents License

### ISSUE THREAD

**Issue by gdi2290** [vuejs/vue#6411]
### What problem does this feature solve?
Flow in under Facebook’s BSD Patents License and this project was built with it

### What does the proposed API look like?
TypeScript which has the Apache License 2.0

**Comment by yyx990803** [comment]
Flow is used only as a dev dependency, and flow syntax is stripped in the built files, so for end users of Vue (e.g. consuming from NPM) there is nothing related to Flow in the distributed code, which doesn't make them affected by Flow's license in any way.

The only concerning case would probably be someone forking Vue and thus using the source code directly - but even so, flow's own code (governed by its license/patent grant) is not distributed in any way either. I doubt that the mere usage of Flow's syntax is subject to its license.

**Comment by nickmessing** [comment]
Considering the fact that typescript team members are interested in Vue (#5887) I would vote for migrating to TypeScript. However, I use flow on day-to-day basics and there is one great feature that I miss in TypeScript (probably more if I do some research). In flow you can use Return type of a function, in TS you cannot.

**Comment by HerringtonDarkholme** [comment]
HI @gdi2290 , I think @yyx990803 has clear reasoning here. Let me explain some technical issue here.

Migrating to TypeScript is doable, but requires many changes at once.

We cannot use `allowJS` option in TS, because TS doesn't accept type annotation in JS as flow does. That means we will need to migrate in one single shot, including building system changes.

TS and Flow still have many subtle syntax differences. These differences are small, but ubiquitous. e.g. https://github.com/vuejs/vue/blob/dev/flow/compiler.js#L4 `?string` in flow is `string | undefined | null` in TS.

Other potential  issues might be TS/Flow's checking rules. e.g. Flow allows `if (o.nonExist) { use(o.nonExist) }` while TS does not. I don't know how much code will be impacted until trying to migrate.

**Comment by gdi2290** [comment]
I created an issue in Flow and they say it's under Facebook’s BSD+Patents License
https://github.com/facebook/flow/issues/4661

most of the syntax for TypeScript and Flow are similar and in some cases TypeScript allows you to be more strict. If the TypeScript is interested in vue then perhaps vue can request some features in Flow that @nickmessing mentioned

**Comment by nickmessing** [comment]
@gdi2290, All features I could find are already requested in the issues but looks like TypeScript team is really busy and they need to clear that 2.5k issues backlog :)

**Comment by kirbyfan64** [comment]
Forgive the random outside opinion, but would this really be worth it? Seems like the only benefit would be a potential licensing "problem" that *doesn't really affect anything outside of Vue itself*...

**Comment by nickmessing** [comment]
@kirbyfan64, as @HerringtonDarkholme said, it's obviously too much human effort and need to be done during a big code freeze and most probably we don't have that much human effort available, but this is still open for discussion.

**Comment by manisenkov** [comment]
@yyx990803 
> ...so for end users of Vue (e.g. consuming from NPM) there is nothing related to Flow in the distributed code

Actually whole `src` folder is distributed with NPM package as it is
https://github.com/vuejs/vue/blob/82f03de3328abde70493d2d3630137f23f140fcb/package.json#L10
However, not sure if this could be treated as something Flow-related for the end user.

**Comment by LinusBorg** [comment]
I would be suprised to find that Flow has any patent on their type annotation *syntax*.

And since our using neither flow nor the vue source files, you are not *using* anything flow-related anyway.

**Comment by russellr86** [comment]
@LinusBorg Flow's license is BSD with Patents, same as React: 

> The license granted hereunder will terminate, automatically and without notice,
> if you (or any of your subsidiaries, corporate affiliates or agents) initiate
> directly or indirectly, or take a direct financial interest in, any Patent
> Assertion: (i) against Facebook or any of its subsidiaries or corporate
> affiliates, (ii) against any party if such Patent Assertion arises in whole or
> in part from any software, technology, product or service of Facebook or any of
> its subsidiaries or corporate affiliates, or (iii) against any party relating
> to the Software. Notwithstanding the foregoing, if Facebook or any of its
> subsidiaries or corporate affiliates files a lawsuit alleging patent
> infringement against you in the first instance, and you respond by filing a
> patent infringement counterclaim in that lawsuit against that party that is
> unrelated to the Software, the license granted hereunder will not terminate
> under section (i) of this paragraph due to such counterclaim.

Which in my opinion, means that any proprietary product you develop, Facebook can come in and develop the same thing without any fear of lawsuit.  Because if you do file suit and you are using Flow (or React or any other BSD+Patent licensed Facebook software on GitHub) your license to use the software terminates, invalidating any proprietary claims you may have had.

TL;DR; yes, I think using the Flow syntax alone does impact Vue.js from a legal standpoint, but I'm not a lawyer, this is just my opinion.

**Comment by LinusBorg** [comment]
> Because if you do file suit and you are using Flow (or React or any other BSD+Patent licensed Facebook software on GitHub) your license to use the software terminates, invalidating any proprietary claims you may have had.

But you are not using Flow, are you? nothing in Vue's code *that you use* contains any hint of Flow.

The fact that *we (the Vue maintainers)* used Flow in our build process to create these files, all *while we had/have a licence to Flow*, is of no concern to you. You are not using Flow or any of it's patented technologies (if there are any, no one could find applications so far).

**Comment by russellr86** [comment]
@LinusBorg, you may be right, like I said I'm not a lawyer or an expert in law.  I just think their licensing goes against the general nature of open source, it's dumb that these types of threads and considerations even need to exist.

**Comment by corbinu** [comment]
I will say if we do decide to convert over I would be happy to help have had a lot of success/practice lately converting projects over and I helping as much as I can on the eslint TypeScript compatibility work. I would be for this for a lot of reason including the patent issues but understand if the majority/core team is not.

**Comment by sandangel** [comment]
How about Vue 3.0 with rewriting in TS and has full support TS like Angular?

**Comment by gdi2290** [comment]
I'm down for Vue 3.0 being TypeScript then we can take learnings from Polymer 3
https://github.com/Polymer/polymer/blob/3.0-preview/lib/utils/async.js

**Comment by amritk** [comment]
I would be interested in helping convert to typescript as well if needed.

**Comment by khaled-nuur** [comment]
doesn't it causes longer learning curve + more Angular-like nature?
I am

**Comment by gdi2290** [comment]
@khaled-nuur typescript would only be used for vue's source code similar to Flow which doesn't mean you need to use it. You still need to learn about types with Flow just as you would with TypeScript the difference is that one has more support and a bigger community

**Comment by corbinu** [comment]
@khaled-nuur Also my understanding is that Angular really tries to enforce full typing when you use TypeScript which to me is pretty silly as I feel like its ability to choose when and where to type is one of its strengths. For example I don't tend to type my application code much but when I break something into a module or component I type it just as I document it. That way 5 months from now when I make an application change I don't need to go reread that component. If that makes any sense. My understanding is Angular kinda more forces you into all or nothing

**Comment by gdi2290** [comment]
Here's an article about how ["JavaScript libraries should be written in TypeScript"](https://staltz.com/all-js-libraries-should-be-authored-in-typescript.html). Just because the codebase is TypeScript doesn't mean the application developer needs to use TypeScript. Angular is more or less targeting enterprise developers which largely benefit from a TypeScript codebase. They also doubled down on decorators which looks better in TypeScript compared to ES6 so their code examples would always look better in TypeScript. They also support ES6 and ES5 but don't really advertise it much

```js
// Angular es5 syntax
(function (app) {
    const AppComponent = ng.core
        .Component({
            selector: "my-app",
            template: "<h1>My First Angular 2 App</h1>"
        })
        .Class({
            constructor: function constructor() {}
        });

    app.AppModule = ng.core.NgModule({
        imports: [ng.platformBrowser.BrowserModule],
        declarations: [AppComponent],
        bootstrap: [AppComponent]
    }).Class({
        constructor: function constructor() {}
    });
}(window.app || (window.app = {})));
```
vs typescript
```typescript
@Component({
  selector: "my-app",
  template: "<h1>My First Angular 2 App</h1>"
})
class AppComponent {
}

@NgModule({
  imports: [BrowserModule],
  declarations: [AppComponent],
  bootstrap: [AppComponent]
})
class AppModule {}
```

Vue's codebase doesn't really mention that it's using a typed language so all of problems people are mentioning for TypeScript in this repo are already there with Flow. Even if Vue uses TypeScript they don't need to use decorators. If Vue did the API would look the same for example

https://github.com/kaorun343/vue-property-decorator
https://github.com/vuejs/vue-class-component

**Comment by mroch** [comment]
we're going to be relicensing Flow under MIT: https://code.facebook.com/posts/300798627056246

hope this relieves the concerns here!

**Comment by yyx990803** [comment]
@mroch wow, big news! I guess this can be closed for now then ;)
